Added a mock temperature sensor to be able to implement the basic fun…

wvreeven opened the pull request:
…ctionality of the CSC. Implemented that basic functionality as well.

pkubanek commented on python/lsst/ts/ess/mock/ess_temperature_4ch_c.py:
```diff
@@ -0,0 +1,34 @@
+# This file is part of ts_Dome.
```
This isn't a mock class. This is class holding temperature 4ch data. Move it out of mock directory please.

wvreeven replied:
You are right that it strictly speaking isn't a mock class. The ESS CSC has been added to replace the current implementation using SALPY, which does contain a EssTemperature4ChC class. I added this to be able to implement a common architecture and I need to discuss the EssTemperature4ChC class with the external developer and once that has been done we will probably introduce a better solution. Still, I have moved this class to a new data directory.

wvreeven replied:
Also, the code isn't part of ts_Dome but of ts_ess (silly copy and paste mistake) so I have corrected that in all Python files that had ts_Dome.

pkubanek commented on python/lsst/ts/ess/mock/mock_temperature_sensor.py:
```diff
@@ -0,0 +1,61 @@
+        await self.determine_current_tai()
+        if self.current_tai - self.previous_tai > 1.0:
+            self.myData.timestamp = self.current_tai
+            self.myData.TemperatureC01 = 20.0
```
I would rather implement 20.0 + random number (such as random.gauss(20.0, 3)?). Same for following channels. This makes EUI/GUIs more happy displaying change in values.

wvreeven replied:
Great suggestion. Done.

pkubanek commented on setup.cfg:
```diff
@@ -1,5 +1,6 @@
 [flake8]
```
Project is shifting towards black formatting. I would recommend the same here, skipping flake8.

wvreeven replied:
I agree that it would be good to start using black (which already is done via a githook) but I prefer to do that in a separate ticket.

pkubanek commented on tests/test_mock_temperature_sensor.py:
```diff
@@ -0,0 +1,47 @@
+        self._ess_sensor.current_tai = salobj.current_tai()
+        await self._ess_sensor.readInstrument()
+        data = self._ess_sensor.myData
+        self.assertEqual(self._ess_sensor.current_tai, data.timestamp)
```
Temperature check here would me more than welcomed. As noted above, I would rather shift to use random numbers for temp reporting. So range check (self.assertEqual(15 <= data.TemperatureC01 <= 25) would be great.

wvreeven replied:
Apart from the fact that in this case self.assertTrue should be used, great suggestion! Done.

pkubanek commented on python/lsst/ts/ess/ess_csc.py:
```diff
@@ -60,9 +152,7 @@ def get_config_pkg():
-        parser.add_argument(
-            "-s", "--simulate", action="store_true", help="Run in simuation mode?"
-        )
+        parser.add_argument("-s", "--simulate", action="store_true", help="Run in simuation mode?")
```
See below. This is black vs. flake8 formatting. I would rather drop this  change and switch to black formatting.

wvreeven replied:
Yes I still agree and as mentioned above I'd like to do that in a separate ticket. And as stated as before, the code was formatted by black via a githook and an IDE plugin and the line length used to be set to 80 and now was changed to 110 so that's why these lines were automatically reformatted.

pkubanek commented on python/lsst/ts/ess/ess_csc.py:
```diff
@@ -36,21 +59,90 @@ def __init__(
+                    previous_tai = data.timestamp
+                    telemetry = {}
+                    telemetry["timestamp"] = data.timestamp
+                    telemetry["TemperatureC01"] = data.TemperatureC01
```
Switching TemeratureC to array will allow for a for loop setting those value (

```python
for in in range(1,5):
    telemetry[f"TemperatureC{i:02}"] = data.temperatureC[i]

wvreeven replied:
I agree that that would be better but I don't know what the internal structure of this class is so until I know that I cannot change this code. I am merely following what's done in the ESS_Temperature4Ch_Publisher.sh file. And I will work with our external developer to improve the code in this respect, including camel casing of the variable names.

pkubanek commented on python/lsst/ts/ess/mock/ess_temperature_4ch_c.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __init__(self):
+        self.timestamp = None
+        self.TemperatureC01 = None
```
Consider changing that to array. Also please use lover case start for variables (temperatureC01) as upper cases are reserved for classes.

wvreeven replied:
I agree that that would be better but I don't know what the internal structure of this class is so until I know that I cannot change this code. I am merely following what's done in the ESS_Temperature4Ch_Publisher.sh file. And I will work with our external developer to improve the code in this respect, including camel casing of the variable names.

wvreeven commented on python/lsst/ts/ess/ess_csc.py:
```diff
@@ -36,21 +59,90 @@ def __init__(
-            raise salobj.ExpectedError(
-                f"Simulation_mode={simulation_mode} must be 0 or 1"
-            )
+            raise salobj.ExpectedError(f"Simulation_mode={simulation_mode} must be 0 or 1")
```
The setting of simulation_mode is done in the add_kwargs_from_args method a bit further down. Implementing both the implement_simulation_mode and add_kwargs_from_args is a pattern followed by many CSCs so I prefer to keep it this way.

wvreeven commented:
I addressed all of your concerns. As you can see, I made several changes and unfortunately cannot comply to all of your suggested improvements. Please have another look taking that into account. Many thanks in advance!

pkubanek commented on python/lsst/ts/ess/ess_csc.py:
```diff
@@ -36,21 +59,85 @@ def __init__(
+                if data.timestamp - previous_tai > 1.0:
+                    previous_tai = data.timestamp
+                    telemetry = vars(data)
+                    self.tel_temperature4Ch.set_put(**telemetry)
```
you can condense that to:

```python
self.tel_temperature4Ch.set_put(**vars(data))
```

wvreeven replied:
Done.

pkubanek commented on python/lsst/ts/ess/data/ess_temperature_4ch_c.py:
```diff
@@ -0,0 +1,34 @@
+
+    def __init__(self):
+        self.timestamp = None
+        self.TemperatureC01 = None
```
temperatureC01 is lower case in XML, shall be lower case here as well

wvreeven replied:
That's correct. As I explained before, the current SALPY EssTemperature4ChC class has it with an upper case T and I cannot break that API. The fact that it should be with a lower case t will be addressed in the near future.
